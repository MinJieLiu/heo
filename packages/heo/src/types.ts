import type { ReactNode } from 'react';

export type SelectorFn<Value, Selected> = (value: Value) => Selected;

export interface StoreProviderProps<State = void, Value = void> {
  initialState?: State;
  children?: ReactNode | ((value: Value) => ReactNode);
}

export interface StoreBridge<Value> {
  value: Value;
  listeners: Set<(value: Value) => void>;
}
