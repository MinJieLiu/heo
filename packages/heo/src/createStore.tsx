/* eslint-disable no-param-reassign,react/display-name */
import React, {
  createContext,
  memo,
  useCallback,
  useContext,
  useEffect,
  useLayoutEffect,
  useMemo,
  useRef,
} from 'react';
import ReactDOM from 'react-dom';
import { useSyncExternalStore } from 'use-sync-external-store/shim';
import { shallow } from './shallow';
import type { StoreBridge, SelectorFn, StoreProviderProps } from './types';

const batchedUpdates = ReactDOM.unstable_batchedUpdates || ((fn: () => void) => fn());

const useIsomorphicLayoutEffect = typeof window === 'undefined' ? useEffect : useLayoutEffect;

/**
 * Create a store with `useHook`
 */
export function createStore<Value extends object, State = void>(
  useHook: (initialState?: State) => Value,
) {
  // Keep the Context never triggering an update
  const Context = createContext<StoreBridge<Value> | null>(null);

  const Provider = memo(({ initialState, children }: StoreProviderProps<State, Value>) => {
    const value = useHook(initialState);
    const bridge = useRef<StoreBridge<Value>>({ value, listeners: new Set() }).current;
    bridge.value = value;

    useIsomorphicLayoutEffect(() =>
      batchedUpdates(() => bridge.listeners.forEach((listener) => listener(value))),
    );

    return (
      <Context.Provider value={bridge}>
        {typeof children === 'function' ? children(value) : children}
      </Context.Provider>
    );
  });

  function useContextValue() {
    const value = useContext(Context);
    if (value === null) {
      throw new Error('The component must be wrapped by <Store.Provider />');
    }
    return value;
  }

  function useStoreUpdate(bridge: StoreBridge<Value>, equalityFn: (nextValue: Value) => boolean) {
    const fnRef = useRef(equalityFn);
    fnRef.current = equalityFn;

    const subscribe = useCallback((update: () => void) => {
      const { listeners } = bridge;
      function listener(nextValue: Value) {
        if (!fnRef.current(nextValue)) {
          update();
        }
      }
      listeners.add(listener);
      return () => listeners.delete(listener);
    }, []);

    const getSnapshot = useCallback(() => bridge.value, []);
    useSyncExternalStore(subscribe, getSnapshot);
  }

  function useSelector<Selected>(selector: SelectorFn<Value, Selected>): Selected {
    const bridge = useContextValue();
    const selected = selector(bridge.value);

    useStoreUpdate(bridge, (nextValue) => shallow(selected, selector(nextValue)));
    return selected;
  }

  function usePicker(): Value {
    const bridge = useContextValue();

    const proxy = useRef<Value>();
    const canUpdate = useRef(false);
    const updated = useRef(false);

    useMemo(() => {
      const handler: ProxyHandler<Value> = {
        get(target, key) {
          canUpdate.current = true;
          target[key] = bridge.value[key];
          return target[key];
        },
        set(target, key, val) {
          if (key in target && val !== target[key]) {
            updated.current = true;
            target[key] = val;
          }
          return true;
        },
      };

      proxy.current = new Proxy({} as Value, handler);
    }, []);

    useStoreUpdate(bridge, (nextValue) => {
      if (!canUpdate.current) return true;
      // Trigger update
      Object.assign(proxy.current as Value, nextValue);
      const result = !updated.current;
      updated.current = false;

      return result;
    });
    return proxy.current as Value;
  }

  function withProvider<T extends object>(
    Children: React.FC<T>,
    initialState?: State,
  ): React.FC<T> {
    return function ProviderComponent(props) {
      return (
        <Provider initialState={initialState}>
          <Children {...props} />
        </Provider>
      );
    };
  }

  return {
    Context,
    Provider,
    useSelector,
    usePicker,
    withProvider,
  };
}
