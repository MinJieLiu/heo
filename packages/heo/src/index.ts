export * from './createStore';
export * from './shallow';
export * from './types';
