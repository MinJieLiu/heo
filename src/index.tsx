import React from 'react';

export type SelectorFn<Value, Selected> = (value: Value) => Selected;

export interface ContainerProviderProps<State = void, Value = void> {
  initialState?: State;
  children?: React.ReactNode | ((value: Value) => React.ReactNode);
}

interface KeepValue<Value> {
  value: Value;
  listeners: Set<(value: Value) => void>;
}

const isSSR =
  typeof window === 'undefined' || /ServerSideRendering/.test(window.navigator && window.navigator.userAgent);

const useIsomorphicLayoutEffect = isSSR ? React.useEffect : React.useLayoutEffect;

/**
 * Create a container with `useHook`
 */
export function createContainer<Value, State = void>(useHook: (initialState?: State) => Value) {
  // Keep the Context never triggering an update
  const Context = React.createContext<KeepValue<Value> | null>(null);

  const Provider = React.memo(({ initialState, children }: ContainerProviderProps<State, Value>) => {
    const value = useHook(initialState);
    const keepValue = React.useRef<KeepValue<Value>>({ value, listeners: new Set() }).current;
    keepValue.value = value;

    if (process.env.NODE_ENV !== 'production') {
      useIsomorphicLayoutEffect(() => {
        keepValue.listeners.forEach((listener) => {
          listener(value);
        });
      });
    } else {
      keepValue.listeners.forEach((listener) => {
        listener(value);
      });
    }

    return (
      <Context.Provider value={keepValue}>
        {typeof children === 'function' ? children(value) : children}
      </Context.Provider>
    );
  });

  function useSelector<Selected>(selector: SelectorFn<Value, Selected>): Selected {
    const [, forceUpdate] = React.useReducer((c) => c + 1, 0);
    const forwardValue = React.useContext(Context);
    if (forwardValue === null) {
      throw new Error('The component must be wrapped by <Container.Provider />');
    }
    const { value, listeners } = forwardValue;
    const selected = selector(value);

    const storeValue = {
      selector,
      value,
      selected,
    };
    const ref = React.useRef(storeValue);
    ref.current = storeValue;

    useIsomorphicLayoutEffect(() => {
      // Trigger update conditions, the same will prevent render
      function callback(nextValue: Value) {
        try {
          if (!ref.current) {
            return;
          }
          const refValue = ref.current;
          if (refValue.value === nextValue) {
            return;
          }
          const nextSelected = refValue.selector(nextValue);
          if (isShadowEqual(refValue.selected, nextSelected)) {
            return;
          }
        } catch (e) {
          // ignore
        }
        forceUpdate();
      }
      listeners.add(callback);
      return () => {
        listeners.delete(callback);
      };
    }, []);
    return selected;
  }

  function usePicker<Selected extends keyof Value>(selected: Selected[]): Pick<Value, Selected> {
    return useSelector((state) => pick(state as Required<Value>, selected));
  }

  function withPicker<T, Selected extends keyof Value>(
    Child: React.FC<T & Pick<Value, Selected>>,
    selected: Selected[],
  ): React.FC<Omit<T, Selected>> {
    return (props) => {
      const picked = usePicker(selected);
      return <Child {...picked} {...(props as T)} />;
    };
  }

  function withProvider<T>(Child: React.FC<T>, initialState?: State): React.FC<T> {
    return (props) => (
      <Provider initialState={initialState}>
        <Child {...props} />
      </Provider>
    );
  }

  return {
    Context,
    Provider,
    /**
     * Filter the required return value through the hook.
     * @param selector
     * @example
     * const count = useSelector(state => state.count)
     */
    useSelector,
    /**
     * Filter the required values in the array, provided that `state` must be `object`.
     * @param selected
     * @example
     * const { count, setCount } = usePicker(['count', 'setCount'])
     */
    usePicker,
    /**
     * High-level component picker
     * @param Child
     * @param selected
     */
    withPicker,
    /**
     * Provider wrapper.
     * @param Child
     */
    withProvider,
  };
}

function pick<T extends object, U extends keyof T>(origin: T, keys: U[]): Pick<T, U> {
  const empty = {} as Pick<T, U>;
  if (!origin) {
    return empty;
  }
  return Object.assign(empty, ...keys.map((key) => ({ [key]: origin[key] })));
}

/**
 * Shadow equal
 * @param origin
 * @param next
 */
function isShadowEqual(origin: unknown, next: unknown) {
  if (Object.is(origin, next)) {
    return true;
  }
  if (origin && typeof origin === 'object' && next && typeof next === 'object') {
    if (
      [...Object.keys(origin), ...Object.keys(next)].every(
        (k) => origin[k] === next[k] && origin.hasOwnProperty(k) && next.hasOwnProperty(k),
      )
    ) {
      return true;
    }
  }
  return false;
}

export { isShadowEqual };

export { default as useFunction } from './useFunction';
export { default as useMethods } from './useMethods';
