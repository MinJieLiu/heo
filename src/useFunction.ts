import React from 'react';

function useFunction<T extends (...args: any[]) => any>(fn: T) {
  const ref = React.useRef<Function>(() => {
    throw new Error('Cannot call function while rendering.');
  });

  ref.current = fn;

  return React.useCallback(ref.current as T, [ref]);
}

export default useFunction;
