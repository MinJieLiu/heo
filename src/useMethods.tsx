import React from 'react';

/**
 * Hook of persistent methods
 */
export default function useMethods<T extends Record<string, (...args: any[]) => any>>(methods: T) {
  const { current } = React.useRef({
    methods,
    func: undefined as T | undefined,
  });
  current.methods = methods;

  if (!current.func) {
    const func = Object.create(null);
    Object.keys(methods).forEach((key) => {
      func[key] = (...args: unknown[]) => current.methods[key].call(current.methods, ...args);
    });
    current.func = func;
  }

  return current.func as T;
}
