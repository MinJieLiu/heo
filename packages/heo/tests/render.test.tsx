import React from 'react';
import { render, fireEvent, cleanup } from '@testing-library/react';
import { createStore } from '../src';

describe('render spec', () => {
  afterEach(cleanup);

  it('should not render with other', () => {
    const CounterContainer = createStore(() => {
      const [count, setCount] = React.useState(0);
      const [total, increment] = React.useReducer((c) => c + 1, 0);

      const [user, setUser] = React.useState({ id: 'jack' });

      const setUserTonny = React.useCallback(() => {
        setUser({ id: 'tonny' });
      }, []);

      const setUserMarry = React.useCallback(() => {
        setUser({ id: 'marry' });
      }, []);

      return {
        count,
        setCount,
        total,
        increment,

        user,
        setUserTonny,
        setUserMarry,
      };
    });

    const Counter1 = React.memo(() => {
      const { count, setCount } = CounterContainer.usePicker();
      const increment = () => setCount((s) => s + 1);

      const renderCount = React.useRef(0);
      if (renderCount.current !== count) throw new Error('Inconsistent rendering (1)!');
      renderCount.current += 1;

      return (
        <div>
          <span>{count}</span>
          <button type="button" onClick={increment}>
            ADD1
          </button>
        </div>
      );
    });

    const Counter2 = React.memo(() => {
      const { total, increment } = CounterContainer.usePicker();

      const renderCount = React.useRef(0);
      if (renderCount.current !== total) throw new Error('Inconsistent rendering (2)!');
      renderCount.current += 1;

      return (
        <div>
          <span>{total}</span>
          <button type="button" onClick={increment}>
            ADD2
          </button>
        </div>
      );
    });

    const Counter3 = React.memo(() => {
      const total = CounterContainer.useSelector((c) => c.total);

      const renderCount = React.useRef(0);
      if (renderCount.current !== total) throw new Error('Inconsistent rendering (3)!');
      renderCount.current += 1;

      return null;
    });

    const Counter4 = React.memo(() => {
      const increment = CounterContainer.useSelector((c) => c.increment);
      return (
        <button type="button" onClick={increment}>
          ADD3
        </button>
      );
    });

    const Counter5 = React.memo(() => {
      const { setUserTonny, setUserMarry } = CounterContainer.usePicker();

      const renderCount = React.useRef(0);
      if (renderCount.current > 0) throw new Error('Inconsistent rendering (5)!');
      renderCount.current += 1;

      return (
        <>
          <button type="button" onClick={setUserTonny}>
            ADD4
          </button>
          <button type="button" onClick={setUserMarry}>
            ADD5
          </button>
        </>
      );
    });

    const Counter6 = React.memo(() => {
      const user = CounterContainer.useSelector((c) => c.user);

      const userChangeCount = React.useRef(0);

      React.useMemo(() => {
        userChangeCount.current += 1;
      }, [user]);

      const renderCount = React.useRef(0);
      renderCount.current += 1;

      if (renderCount.current > userChangeCount.current)
        throw new Error('Inconsistent rendering (6)!');

      return <div>{user.id}</div>;
    });

    const App = CounterContainer.withProvider(() => (
      <>
        <Counter1 />
        <Counter2 />
        <Counter3 />
        <Counter4 />
        <Counter5 />
        <Counter6 />
      </>
    ));
    const { getAllByText, container } = render(<App />);
    expect(container).toMatchSnapshot();

    expect(() => fireEvent.click(getAllByText('ADD1')[0])).not.toThrow();
    expect(() => fireEvent.click(getAllByText('ADD2')[0])).not.toThrow();
    expect(() => fireEvent.click(getAllByText('ADD3')[0])).not.toThrow();
    expect(() => fireEvent.click(getAllByText('ADD4')[0])).not.toThrow();
    expect(() => fireEvent.click(getAllByText('ADD5')[0])).not.toThrow();

    expect(container).toMatchSnapshot();
  });
});
