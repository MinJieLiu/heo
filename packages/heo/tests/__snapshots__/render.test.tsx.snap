// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render spec should not render with other 1`] = `
<div>
  <div>
    <span>
      0
    </span>
    <button
      type="button"
    >
      ADD1
    </button>
  </div>
  <div>
    <span>
      0
    </span>
    <button
      type="button"
    >
      ADD2
    </button>
  </div>
  <button
    type="button"
  >
    ADD3
  </button>
  <button
    type="button"
  >
    ADD4
  </button>
  <button
    type="button"
  >
    ADD5
  </button>
  <div>
    jack
  </div>
</div>
`;

exports[`render spec should not render with other 2`] = `
<div>
  <div>
    <span>
      1
    </span>
    <button
      type="button"
    >
      ADD1
    </button>
  </div>
  <div>
    <span>
      2
    </span>
    <button
      type="button"
    >
      ADD2
    </button>
  </div>
  <button
    type="button"
  >
    ADD3
  </button>
  <button
    type="button"
  >
    ADD4
  </button>
  <button
    type="button"
  >
    ADD5
  </button>
  <div>
    marry
  </div>
</div>
`;
